Added tests for FENothing constraints.

marcfehling opened the pull request:
~~Blocked by #11447.~~ (EDIT: Has been merged)

The updated documention in #11430 states that `FE_Nothing` elements impose continuity constraints on neighboring continuous elements when their `dominate` parameter is set to `true`.

However, this test states that this is not the case: the `dominate` parameter has no effect. I feel like I stumbled over a bug.

@bangerth -- FYI

May be related to #1496.

bangerth commented on tests/hp/fe_nothing_23.cc:
```diff
@@ -0,0 +1,115 @@
+  initlog();
+
+  deallog << std::boolalpha;
+  for (const bool fe_nothing_dominates : {false, true})
```
I think it's just cool that C++11 allows for this construct :-)

bangerth reviewed (commented):
Yes, this looks like a bug. I guess the `constraints` object is empty in all cases? What happens if you exchange which element is on the two cells (in other words, from which side we end up looking at the interface)?

My suspicion is that the face interpolation matrix returned when looking at that face is `Nx0` and so we end up not enforcing any constraints. I'm confused why we didn't notice that when @davydden added that flag...

marcfehling commented:
> I guess the `constraints` object is empty in all cases?

Yes, the `constraints` object is empty in all cases.

> What happens if you exchange which element is on the two cells (in other words, from which side we end up looking at the interface)?

The result is the same. The `constraints` object is still empty. And all dofs on the FEQ cell all have the value 1.

marcfehling commented:
In `DoFTools::make_hp_hanging_node_constraints()`, we always add constraints from the point of view of the dominating finite element. However, we break if there are no face dofs, and thus no constraints will be imposed if `FENothing` elements dominate.
https://github.com/dealii/dealii/blob/b985807f268b6e179482ea06776f0c4ba75e305c/source/dofs/dof_tools_constraints.cc#L1509-L1513

I do not like that the `FENothing` class needs special treatment in so many different parts in the library, but I guess there is no other way here. I would replace the `break` statement and constrain all `dependent_dofs` to the value `0` here. Maybe check with a dynamic cast whether the neighboring FE is actually `FENothing`, as there may be other finite elements without face dofs.

Do you have a different (better) idea here?

marcfehling commented:
Let us merge this one too according to https://github.com/dealii/dealii/pull/11549#pullrequestreview-572817961.

marcfehling commented:
Since #11448 #11549 and #11600 basically did the same, I moved them into this PR, and moved the core function into a separate header file. Ready for review and tests.

bangerth commented:
/rebuild
